driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java

```diff
@@ -0,0 +1,544 @@
+ * <p>
+ * For more flexibility, the query logger uses 3 different {@link Logger} instances:
+ *
+ * <ol>
+ *     <li>{@link #NORMAL_LOGGER}: used to log normal queries, i.e., queries that completed successfully
+ *     within a {@link #slowQueryLatencyThresholdMillis configurable threshold in milliseconds}.</li>
+ *     <li>{@link #SLOW_LOGGER}: used to log slow queries, i.e., queries that completed successfully
+ *     but that took longer than a {@link #slowQueryLatencyThresholdMillis configurable threshold in milliseconds} to complete.</li>
+ *     <li>{@link #ERROR_LOGGER}: used to log unsuccessful queries, i.e.,
+ *     queries that did not completed normally and threw an exception.
+ *     Note this this logger will also print the full stack trace of the reported exception.</li>
+ * </ol>
+ *
+ * <p>
+ * The appropriate logger is chosen according to the following algorithm:
+ * <ol>
+ *     <li>if an exception has been thrown: use {@link #ERROR_LOGGER};</li>
+ *     <li>otherwise, if the reported latency is greater than {@link #slowQueryLatencyThresholdMillis the configured threshold in milliseconds}: use {@link #SLOW_LOGGER};</li>
+ *     <li>otherwise, use {@link #NORMAL_LOGGER}.</li>
+ * </ol>
+ *
+ * <p>
+ * All loggers are activated by setting their levels to {@code DEBUG} or {@code TRACE} (including {@link #ERROR_LOGGER}).
+ * If the level is set to {@code TRACE} and the statement being logged is a {@link BoundStatement},
+ * then the query parameters (if any) will be logged as well (names and actual values).
+ *
+ * <p>
+ * This class is thread-safe.
+ *
+ * @since 2.0.10
+ */
+public class QueryLogger implements LatencyTracker {
+
+    /**
+     * The default threshold in milliseconds beyond which queries are considered 'slow'
+     * and logged as such by the driver.
+     */
+    public static final long DEFAULT_SLOW_QUERY_THRESHOLD_MS = 5000;
+
+    /**
+     * The default maximum length of a CQL query string that can be logged verbatim
+     * by the driver. Query strings longer than this value will be truncated
+     * when logged.
+     */
+    public static final int DEFAULT_MAX_QUERY_STRING_LENGTH = 500;
+
+    /**
+     * The default maximum length of a query parameter value that can be logged verbatim
+     * by the driver. Parameter values longer than this value will be truncated
+     * when logged.
+     */
+    public static final int DEFAULT_MAX_PARAMETER_VALUE_LENGTH = 50;
+
+    /**
+     * The default maximum number of query parameters that can be logged
+     * by the driver. Queries with a number of parameters higher than this value
+     * will not have all their parameters logged.
+     */
+    public static final int DEFAULT_MAX_LOGGED_PARAMETERS = 50;
+
+    // Loggers
+
+    /**
+     * The logger used to log normal queries, i.e., queries that completed successfully
+     * within at most {@link #DEFAULT_SLOW_QUERY_THRESHOLD_MS} milliseconds.
```

Comment:
Shouldn't this link to slowQueryLatencyThresholdMillis instead of DEFAULT_SLOW_QUERY_THRESHOLD_MS?   Same for SLOW_LOGGER.